Lesson5/Sulima Denis Homework 3/src/functions.cpp

```diff
@@ -1,66 +1,68 @@
 #include "functions.h"
-#include <iostream>
 #include <set>
-#include <string>
 #include <map>
-#include <vector>
+#include <algorithm>
+#include <sstream>
+
+struct cmp
+{
+    bool operator() (std::string lhs, std::string rhs)
+    {
+        std::string tmp_lhs {lhs};
+        std::string tmp_rhs {rhs};
+        std :: transform( tmp_lhs.begin(), tmp_lhs.end(), tmp_lhs.begin(), ::tolower );
+        std :: transform( tmp_rhs.begin(), tmp_rhs.end(), tmp_rhs.begin(), ::tolower );
+        if(tmp_lhs < tmp_rhs)
+            return true;
+        return  false;
+    }
+};
 
 std::vector<std::string> getUniqueWords (const std::string& text){
-    std::set<std::string>textCheck;
-    std::vector<std::string>result;
+    std::set<std::string, cmp>textCheck;
+    std::stringstream textStream;
     std::string tmp;
-    for(const auto &value: text){
-        if(value != ' ')
-            tmp.push_back(tolower(value));
-        else{
-            textCheck.insert(tmp);
-            tmp.clear();
-        }
-    }
-    if(!tmp.empty()){
+    textStream << text;
+    while(!textStream.eof())
+    {
+        textStream >> tmp;
         textCheck.insert(tmp);
     }
-    for(const auto &value: textCheck){
-        result.push_back(value);
-    }
+    std::vector<std::string>result{textCheck.begin(), textCheck.end()};
     return result;
 }
 
+
 std::vector<std::pair<std::string, size_t>> wordCounter (const std::string& text){
-    std::map<std::string, size_t>result;
-    std::vector<std::pair<std::string, size_t>> resultReturn;
-    std::string tmp;
-    for(const auto &value: text){
-        if(value != ' ')
-            tmp.push_back(tolower(value));
+    std::map<std::string, size_t, cmp>result;
+    std::vector<std::pair<std::string, size_t>> resultReturn{};
+    std::string tmp{};
+    if(text == tmp){
+        result.emplace("",0);
+        resultReturn.push_back(*result.begin());
+        return resultReturn;
+    }
+    std::stringstream textStream;
+    textStream << text;
+    while(!textStream.eof())
+    {
+        textStream >> tmp;
+        if(result.empty()){
+            result.emplace(tmp, 1);
+            tmp.clear();
+        }
         else{
-            if(result.empty()){
-                result.emplace(tmp, 1);
+            if(result.find(tmp) != result.end()){
```

Comment:
you should save this iterator, because in another case you'll make two lookups instead of one